### PR TITLE
Add support for adding scope and generate expiry time based on incoming assertion

### PR DESCRIPTION
**Pull request by Megala21**
### Proposed changes in this pull request
- This PR provides support for adding scope with self contained access token and creates expiry time based on incoming expiry time assertion.
- Related issue https://github.com/wso2/product-is/issues/3183
- Related PR https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/892

### When should this PR be merged
N/A

### Follow up actions
N/A

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/JWTTokenIssuer.java [hunk 1]
```diff
@@ -416,6 +427,64 @@ protected JWTClaimsSet createJWTClaimSet(OAuthAuthzReqMessageContext authAuthzRe
...
+        String scopeString = null;
+        if (tokenReqMessageContext != null) {
+            scope = tokenReqMessageContext.getScope();
+        } else {
+            scope = authAuthzReqMessageContext.getApprovedScope();
+        }
+        if (ArrayUtils.isNotEmpty(scope)) {
+            scopeString = OAuth2Util.buildScopeString(scope);
+            if (log.isDebugEnabled()) {
+                log.debug("Scope exist for the jwt access token with subject " + jwtClaimsSet.getSubject() + " and "
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/JWTTokenIssuer.java]: Why not get the subject from the context?
We are building the jwt claimset multiple times unnecessarily just to get context info for logs. IMO we can use information from the context instead

Comment (Megala21) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/JWTTokenIssuer.java]: Changed with the commit 63aa142358c853c194a1d0035fb99fcaf16ac479

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/JWTTokenIssuer.java [hunk 2]
```diff
@@ -416,6 +427,81 @@ protected JWTClaimsSet createJWTClaimSet(OAuthAuthzReqMessageContext authAuthzRe
...
+    private String getAuthenticatedSubjectIdentifier(OAuthAuthzReqMessageContext authAuthzReqMessageContext,
+            OAuthTokenReqMessageContext tokenReqMessageContext) {
+
+        AuthenticatedUser authenticatedUser;
+        if (authAuthzReqMessageContext != null) {
+            authenticatedUser = authAuthzReqMessageContext.getAuthorizationReqDTO().getUser();
+        } else {
+            authenticatedUser = tokenReqMessageContext.getAuthorizedUser();
+        }
+        return authenticatedUser != null ? authenticatedUser.getAuthenticatedSubjectIdentifier() : null;
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/JWTTokenIssuer.java]: In the null case shouldn't we break the flow?
We need to decide whether a self-contained access token with no subject confirmation is acceptable.

Comment (Megala21) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/JWTTokenIssuer.java]: The relevant authenticated user will be set by relevant grant handlers. Hence the resposibility is purely dependant on the grant handlers. Further I moved this code to separate method from previously existing code segment.

Comment (Megala21) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/JWTTokenIssuer.java]: Throwing the exception in the event authenticated user is null

**Pull request merged**